includes/commerce_pos.transaction.inc

```diff
@@ -706,3 +762,89 @@ function commerce_pos_transaction_pay_submit($form, &$form_state) {
   $form_state['transaction']->doAction('save');
   $form_state['redirect'] = 'admin/commerce/pos/' . $form['#redirect_key'] . '/payment';
 }
+
+/**
+ * Validate function for the order search textfield to ensure we have a valid
+ * order number.
+ */
+function _commerce_pos_order_search_validate($form, &$form_state) {
+  if (empty($form_state['values']['order_search']['input'])) {
+    form_set_error('order_search][input', t('Order number cannot be empty.'));
+  }
+  else if (!ctype_digit($form_state['values']['order_search']['input'])) {
+    form_set_error('order_search][input', t('Order number is invalid.'));
+  }
+}
+
+/**
+ * Returns a modal form displaying each item in an order along with a checkbox
+ * for the user to add that item to the return transaction.
+ */
+function commerce_pos_return_select_item_popup(&$form, &$form_state, $order) {
+
+  // Load our generic modal form.
+  $form['commerce_pos_return_select_item_modal'] = commerce_pos_form_modal(array('closeable' => TRUE, 'throbber' => FALSE));
+
+  // Initialize variables.
+  $options = array();
+  $line_items = array();
+
+  // Go through each item in the order and generate the html for the items
+  // display.
+  foreach (entity_metadata_wrapper('commerce_order', $order)->commerce_line_items as $delta => $line_item_wrapper) {
+    if (in_array($line_item_wrapper->type->value(), commerce_product_line_item_types())) {
+
+      // Load the product.
+      $product_sku = $line_item_wrapper->commerce_product->sku->value();
+      if (!empty($product_sku) && $product = commerce_product_load_by_sku($product_sku)) {
+
+        // Add all the items one by one to our markup variable so we can add
+        // that to the checkboxes element below.
+        $markup = '';
+
+        if ($image = _commerce_pos_product_thumbnail($product)) {
+          $markup .= '<div class="commerce-pos-return-product-image">' . $image . '</div>';
+        }
+
+        $markup .= '<div class="commerce-pos-return-product-info"><div class="commerce-pos-return-product-title">' . $product->title_field[LANGUAGE_NONE][0]['safe_value'] . '</div>';
+
+        $markup .= '<div class="commerce-pos-return-product-sku">' . t('SKU: @sku', array('@sku' => $product_sku)) . '</div>';
+
+        // Make sure we display the actual price paid to account for any
+        // discounts the user might have received.
+        $markup .= '<div class="commerce-pos-return-product-price-paid">'
+          . t('Price Paid: @price', array('@price' => commerce_currency_format($line_item_wrapper->commerce_unit_price->value()['amount'], $line_item_wrapper->commerce_unit_price->value()['currency_code'])))
```

Comment:
Note that this "Price Paid" is a unit price and not a total price... Not sure if that matters. But it would to me if I was a customer returning 4 doodads and getting back the money for only one.